Sources/Datadog/Core/HookedURLSession.swift

```diff
@@ -0,0 +1,280 @@
+
+ It intercepts and observes requests and tasks respectively.
+ */
+internal final class HookedSession: NSObject {
+    typealias RequestInterceptor = (URLRequest) -> URLRequest
+    typealias TaskObserver = (URLSessionTask, URLSessionTask.State?) -> Void
+
+    private let session: URLSession
+    let requestInterceptor: RequestInterceptor
+    let taskObserver: TaskObserver
+    private var observations = [Int: NSKeyValueObservation](minimumCapacity: 100)
+
+    init(session: URLSession,
+         requestInterceptor: @escaping RequestInterceptor,
+         taskObserver: @escaping TaskObserver) {
+        self.session = session
+        self.requestInterceptor = requestInterceptor
+        self.taskObserver = taskObserver
+        super.init()
+    }
+
+    func asURLSession() -> URLSession {
+        let castedSession: URLSession = unsafeBitCast(self, to: URLSession.self)
+        return castedSession
+    }
+
+    // MARK: - Transparent messaging
+    /*
+     As a NSObject subclass yet exposed as URLSession (ref: URLSession.tracedSession)
+     all the method calls are `unrecognized_selector` for HookedSession, except
+     those which are implemented in HookedSession.
+     forwardingTarget passes those unimplemented method calls to session
+     */
+    override func forwardingTarget(for aSelector: Selector!) -> Any? { // swiftlint:disable:this implicitly_unwrapped_optional
+        return session
+    }
+
+    // MARK: - Helpers
+
+    /*
+     IMPORTANT NOTE:
+     If you create an URLSessionTask instance from an URLSession instance,
+     the task stays alive EVEN IF you nullify the session instance.
+     This happens because URLSessionTask has private `__taskGroup` property which keeps
+     session instance alive as long as the task is alive.
+
+     This is not the case for HookedSession instances.
+
+     let task = hookedSession.dataTask(...)
+     hookedSession = nil
+     task.resume()
+
+     In the case above, task and
+     hookedSession.session (private property) will stay alive
+     yet hookedSession will be deallocated.
+     That means observations will be deallocated too.
+     In order to keep observing until the task completes,
+     observationBlock below captures `self` on purpose.
+     Therefore, observationBlock keeps `self` alive
+     until the block is removed from self.observations dict.
+     */
+    private func observed<T: URLSessionTask>(_ task: T) -> T {
+        let observer = taskObserver
+        var previousState: URLSessionTask.State? = nil
+        let observation = task.observe(\.state, options: [.initial]) { observed, _ in
+            observer(observed, previousState)
+            previousState = observed.state
+
+            switch observed.state {
+            case .canceling, .completed:
+                self.observations[observed.taskIdentifier] = nil
+            default:
+                break
+            }
+        }
+        observations[task.taskIdentifier] = observation
```

Comment:
Continuing my thought on threading vs completion callbacks, this for instance would require additional synchronisation of `self.observations` as it might be accessed by different threads. Because `TaskObserver` uses `URLSessionTask`, we'd have to pass the `task` to another thread, possibly changing its lifespan. Some client's app might rely on `task` deallocation callback and here we'd be changing this behaviour 💥. This is too invasive IMO, and again reminds me problems we struggled with before #72.